keps/sig-storage/20200212-volume-group.md

```diff
@@ -0,0 +1,1067 @@
+
+## Proposal for Consistency Groups and Group Snapshots
+
+This proposal introduces new CRDs VolumeGroup, VolumeGroupClass, and VolumeGroupSnapshot.
+
+Create new VolumeGroup can be done in several ways:
+1. Create an empty group first, then create a new PVC with the group name which will add a volume to the already created group.
+2. Create an empty group first, and then add an existing PVC to the group one by one.
+3. Create a new volume group from an existing group snapshot.
+4. Create a new volume group and add existing PVCs matching the label selector to the group.
+5. Non-goal: Create a new empty group and in the same time create new empty PVCs and add to the new group.
+
+Modify an existing VolumeGroup:
+Add new volume or remove existing volume from an existing VolumeGroup. Option 2 for create VolumeGroup above falls into this case.
+
+### Create and Modify VolumeGroup
+
+VolumeGroups can be created and/or modified in several ways as described in the following.
+
+#### Create new PVC and add to the VolumeGroup
+
+* Create a new empty VolumeGroup.
+* Create a new PVC with existing VolumeGroup name. As a result, new PVC is created and added to VolumeGroup. VolumeGroup is modified so Status has this new PVC in PVCList.
+* External-provisioner will be modified so that VolumeGroupName will be passed to the CSI driver when creating a volume.
+
+Only CSI drivers supporting VOLUMEGROUP capability can support the volume group this way.
+When a new PVC is created with the existing VolumeGroup name, the VolumeGroup will be modified and the PVC will be added to PVCList in the Status.
+
+#### Modify VolumeGroup with existing PVCs
+
+We can add an existing PVC to the group or remove a PVC from the group without deleting it. A VOLUMEGROUP_ADD_REMOVE_EXISTING_VOLUME capability will be added to CSI Spec. Only CSI drivers supporting both VOLUMEGROUP and VOLUMEGROUP_ADD_REMOVE_EXISTING_VOLUME capabilities can support the volume group this way.
+* Create a new empty VolumeGroup.
+* Add an existing PVC to an existing VolumeGroup (VolumeGroup can be empty to start with or it can have other PVCs already) by adding VolumeGroup name to the PVC Spec.
+* VolumeGroup is modified so the existing PVC is added to the PVCList in the Status.
+  * Note: The VolumeGroup controller will be implemented to have a desired state
+    of the world and an actual state of the world. The desired state of the world
+    contains VolumeGroups with the desired PVCList while the actual state of the
+    world contains VolumeGroups with the actual PVCList. The controller will try
+    to reconcile the two by handling adding and removing multiple PVCs through a
+    single CSI ModifyVolumeGroup RPC call each time.
+* External-provisioner will be modified so that modifying PVC by adding VolumeGroupName will trigger a ModifyVolume call (a new CSI controller RPC) to CSI driver.
```

Comment:
Do we still need ModifyVolume CSI RPC if we have ModifyVolumeGroup RPC?

Reply:
If we need to call ModifyVolume, should that fail the entire modify group volume?
If one volume fails to be added, it should not affect it if it is used by a pod, but there will be error messages.

Reply:
Removed ModifyVolume CSI RPC as we now have ModifyVolumeGroup RPC.